src/dhtproto/client/helper/ChannelSerializer.d

```diff
@@ -0,0 +1,354 @@
+
+    /// Fork helper for asynchronous dumping.
+    private SafeFork fork;
+
+    /// TODO
+    private alias void delegate ( void delegate ( ref hash_t, ref S ) )
+        DumpRecordDg;
+
+    /// TODO
+    private alias void delegate ( hash_t, ref Contiguous!(S) ) LoadRecordDg;
+
+    /// Delegate currently in use by the fork for providing the records to be
+    /// dumped. Only non-null while the fork is running.
+    // private MapSerializer.AdderDg!(hash_t, S) current_record_dump_dg;
+    private DumpRecordDg current_record_dump_dg;
+
+    /// Delegate for providing the records to be dumped, for use by the next
+    /// call of the fork. (`current_record_dump_dg` is only overwritten when
+    /// the fork is not running. For this reason, we need to store the delegate
+    /// to be passed to the *next* call until the current call finishes.)
+    // private MapSerializer.AdderDg!(hash_t, S) new_record_dump_dg;
+    private DumpRecordDg new_record_dump_dg;
+
+    /// Buffer used for de/serialization of record values.
+    private void[] serialize_buf;
+
+    /***************************************************************************
+
+        Constructor.
+
+        Params:
+            path = path of the dump file
+
+    ***************************************************************************/
+
+    public this ( istring path )
+    {
+        const buffer_size = 64 * 1024;
+
+        this.path = path;
+        this.buffered_output = new BufferedOutput(null, buffer_size);
+        this.buffered_input  = new BufferedInput(null, buffer_size);
+        this.fork = new SafeFork(&this.forkEntryPoint);
+    }
+
+    /***************************************************************************
+
+        Dumps the provided associative array to disk.
+
+        Params:
+            aa = associative array to dump
+            block = if true, performs a blocking dump; otherwise performs an
+                asynchronous dump (see module header)
+
+        Returns:
+            true if a dump occurred; false if it was skipped
+
+    ***************************************************************************/
+
+    public bool dump ( S[hash_t] aa, bool block = false )
+    {
+        void dumpAA ( void delegate ( ref hash_t, ref S ) dump_record )
+        {
+            foreach ( k, v; aa)
+                dump_record(k, v);
+        }
+        return this.dump(&dumpAA, block);
+    }
+
+    /***************************************************************************
+
+        Dumps the provided hash map to disk.
+
+        Params:
+            aa = associative array to dump
+            block = if true, performs a blocking dump; otherwise performs an
+                asynchronous dump (see module header)
+
+        Returns:
+            true if a dump occurred; false if it was skipped
+
+    ***************************************************************************/
+
+    public bool dump ( HashMap!(S) map, bool block = false )
```

Comment:
Bear in mind that applications may be using other cache types than built-in AA's or `HashMap` (e.g. `StandardKeyKashingMap`).

Reply:
Ah yes, I was thinking that `StandardKeyKashingMap` was dervied from `HashMap`, but it's not.

Arbitrary types are supported too, by the last overload.

Reply:
Gotcha, makes sense.  Although it might be possible to simplify the different overloads somewhat by implementing a generic mechanism that supports any type that can be `foreach`'d over.

Reply:
The map and AA overloads are purely convenience wrappers around the overloads for generic types.